### DISCUSSION TITLE
GraphQLTestTemplate is not injected at all

### DISCUSSION THREAD

**Discussion by khteh** [graphql-java-kickstart/graphql-spring-boot#419] [Q&A]
**Describe the bug**
I am trying to test my GraphQL query with GraphQLTestTemplate. Here is my test class:
```
@GraphQLTest
public class GraphQLApplicationTests {
	@Autowired
    private GraphQLTestTemplate graphQLTestTemplate;	
	@Test
    public void GetAllBooks_Succeed() throws IOException {
		assertNotNull(graphQLTestTemplate);
        GraphQLResponse response = graphQLTestTemplate.postForResource("graphql/get-all-books.graphql");
        assertNotNull(response);
        assertTrue(response.isOk());
        // bookRepository.save(new Book("My First GrapQL Book", "1234567890", 500, author));
        assertEquals("My First GrapQL Book", response.get("$.data.book.title"));
        assertEquals("1234567890", response.get("$.data.book.isbn"));
        assertEquals(500, response.get("$.data.book.pageCount"));
        assertNotNull("$.data.book.author");
    }
}
```
pom.xml:
```
		<dependency>
		    <groupId>com.graphql-java-kickstart</groupId>
		    <artifactId>graphql-spring-boot-starter-test</artifactId>
		    <version>7.0.1</version>
		</dependency>
```
`mvn test` fails the first assertion: `assertNotNull(graphQLTestTemplate);`

Besides, running it with `mvn spring-boot:run` bumps into the following runtime exception:
```
Error starting ApplicationContext. To display the conditions report re-run your application with 'debug' enabled.
2020-06-10 09:05:24.384 ERROR 28632 --- [           main] o.s.boot.SpringApplication               : Application run failed

org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'graphQLTestUtils' defined in class path resource [com/graphql/spring/boot/test/GraphQLTestAutoConfiguration.class]: Post-processing of merged bean definition failed; nested exception is java.lang.IllegalStateException: Failed to introspect Class [com.graphql.spring.boot.test.GraphQLTestTemplate] from ClassLoader [jdk.internal.loader.ClassLoaders$AppClassLoader@4e0e2f2a]
```

**To Reproduce**

**Expected behavior**

**Screenshots**

**Desktop (please complete the following information):**
 - OS: [e.g. iOS] Windows 10
 - Browser [e.g. chrome, safari] Chrome

**Additional context**

**Comment by BlasiusSecundus** [comment]
Does it work if you are using `@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)` instead of `@GraphQLTest`? (I hardly ever used `@GraphQLTest` to run integration tests).

(Btw. I think the dependency should have `<scope>test</scope>`)

**Comment by khteh** [comment]
No, it is still NULL using `SpringBootTest`.Yes, I missed out the `<scope>test</scope>`

**Comment by BlasiusSecundus** [comment]
At least the spring-boot run issue should have been resolved by adding the test scope.

From the code provided here I have no more ideas, maybe it is related to maven setup?

**Comment by khteh** [comment]
Which maven setup is suspicious of causing the error?

**Comment by BlasiusSecundus** [comment]
That was just an idea, the cause is perhaps something different. I don't have any specific suspicion.

**Comment by jeet23** [comment]
I am facing the same error.
Upon injecting 
``` 
@Autowired
private GraphQLTestTemplate graphqlTestTemplate;
```
IntelliJ complains: `Could not autowire. No beans of 'GraphQLTestTemplate' type found.`.
Upon running the test, I get NullPointerException for graphqlTestTemplate.

```
<dependency>
        <groupId>com.graphql-java-kickstart</groupId>
        <artifactId>graphql-spring-boot-starter-test</artifactId>
        <version>7.0.1</version>
        <scope>test</scope>
</dependency>
```

I tried using `@SpringBootTest` instead of `@GraphQLTest` but still its the same error.

Does anyone have an idea about this?

**Comment by ubarua123** [comment]
Yeah I am facing the same. 
```
@RunWith(SpringRunner::class)
@GraphQLTest
internal class QueryResolverTest {

    @Autowired
    private lateinit var graphQLTestTemplate: GraphQLTestTemplate


    @Test
    fun testFoo() {

    }
}
```

Could not autowire. I noticed there is no `@Component` or `@Bean` annotation on the `GraphQLTestTemplate` class. Maybe that's why? So how is it being declared as a bean.

**Comment by BlasiusSecundus** [comment]
The bean is defined here: https://github.com/graphql-java-kickstart/graphql-spring-boot/blob/master/graphql-spring-boot-test-autoconfigure/src/main/java/com/graphql/spring/boot/test/GraphQLTestAutoConfiguration.java

**Comment by jeet23** [comment]
Ok, so I removed `@GraphQLTest` and I am using 
```
@RunWith(SpringRunner.class)
@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
```
Even though IntelliJ complains, the graphqlTestTemplate bean is actually injected on application startup.

**Comment by Chair4ce** [comment]
I played around with it as well, following @jeet23's comment worked for me. Thanks!


    @Autowired
    private GraphQLTestTemplate graphQLTestTemplate;

    @MockBean
    private ResourceService resourceServiceMock;

    @Test
    public void getResourceById() throws IOException {
        Resource resource = new Resource(1L, "Facebook", "facebook.com");
        when(resourceServiceMock.getResource(any())).thenReturn(resource);
        GraphQLResponse response = graphQLTestTemplate.postForResource("graphql/get-resource-by-id.graphql");
        assertTrue(response.isOk());

        assertEquals("1", response.get("$.data.resource.id"));
        assertEquals("Facebook", response.get("$.data.resource.title"));
        assertEquals("facebook.com", response.get("$.data.resource.url"));
    }
